RSS Block: Add option to open links in new tab/window and control rel attributes

dhruvikpatel18 opened the pull request:
Closes #49285

## What?
This PR adds new options to the RSS block allowing users to:

- Open RSS feed links in a new tab/window
- Add custom rel attributes to links through a simple text input

## Why?
When linking to external websites through an RSS feed, users often want the ability to:
- Open links in a new tab to keep visitors on their site
- Add rel attributes for SEO or linking purposes


## Testing Instructions
1. Add an RSS block to a post or page
2. Enter a valid RSS feed URL
3. In the block sidebar, locate the "Settings" panel
4. Toggle "Open links in new tab" to ON and verify links open in a new tab
5. Add rel attributes like "nofollow sponsored" in the "Link Rel" field in Advanced panel
6. Verify that the HTML output includes all expected attributes

## Screenshots or screencast

https://github.com/user-attachments/assets/e6a9eb15-7cf1-4774-b460-3af953e4b306

dhruvikpatel18 committed:
Add options to open RSS links in new tab and add rel attributes

dhruvikpatel18 committed:
moved link settings in settings panel

Mamaduka commented on packages/block-library/src/rss/block.json:
```diff
@@ -39,6 +39,18 @@
+			"type": "boolean",
+			"default": false
+		},
+		"addNofollow": {
+			"type": "boolean",
+			"default": false
+		},
+		"additionalRelAttributes": {
+			"type": "string",
+			"default": ""
```
I think there's no need to have two separate attributes for `rel`. Maybe we should just display single text control, using a single attribute (`linkRel`) and let people input `nofollow` or any additional field as needed.

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,6 +31,32 @@ function render_block_core_rss( $attributes ) {
+
+	if ( ! empty( $attributes['openInNewTab'] ) ) {
+		$link_attributes .= ' target="_blank"';
+		$rel_attributes[] = 'noopener noreferrer';
```
As far as I know, these values are no longer required. See related discussion in https://github.com/WordPress/gutenberg/issues/26914.

dhruvikpatel18 committed:
Simplify link attributes with single rel input

dhruvikpatel18 commented:
@Mamaduka Thank you for the feedback! 

I've updated the implementation to address your suggestions:

- Consolidated the rel attributes into a single `linkRel` attribute
- Removed the separate `addNofollow` boolean
- Simplified the input to a single text control for rel attributes
- Removed automatic `noopener noreferrer` attributes as suggested

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,6 +31,28 @@ function render_block_core_rss( $attributes ) {
+		}
+	}
+
+	if ( ! empty( $rel_attributes ) ) {
+		$link_attributes .= ' rel="' . esc_attr( implode( ' ', array_unique( $rel_attributes ) ) ) . '"';
+	}
```
I think this is unnecessary; you're already assigning value above.

PS There're some WPCS violations in this file that also need to be fixed.

Mamaduka commented on packages/block-library/src/rss/block.json:
```diff
@@ -39,6 +39,14 @@
+			"type": "boolean",
+			"default": false
+		},
+		"linkRel": {
```
```suggestion
		"rel": {
```

I just remembered that the Social Icons block also has a custom rel handler; let's make it consistent.

https://github.com/WordPress/gutenberg/blob/bb9028dc46c8422c9319c7d04149ad21312e30d0/packages/block-library/src/social-link/block.json#L22-L24

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,6 +31,28 @@ function render_block_core_rss( $attributes ) {
+
+
+	if ( ! empty( $attributes['linkRel'] ) ) {
+		$rel_attributes = explode( ' ', $attributes['linkRel'] );
+		$rel_attributes = array_filter( array_map( 'sanitize_html_class', $rel_attributes ) );
+
+		if ( ! empty( $rel_attributes ) ) {
+			$link_attributes .= ' rel="' . esc_attr( implode( ' ', $rel_attributes ) ) . '"';
+		}
```
Do we really need all this sanitization for escaped value? Could this be something like:

```php
$link_attributes .= ' rel="' . esc_attr( trim( $attributes['rel'] ) ) . '"';
```

Mamaduka reviewed (changes requested):
Thank you, @dhruvikpatel18!

I just remembered that the Social Icon block also has a custom "Rel" field, let's match UI and implementation (minus HTML API) to it.

P.S. Rebasing should resolve failing e2e tests.

dhruvikpatel18 committed:
Refactored new tab and rel attribute handling

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -39,9 +43,18 @@ function render_block_core_rss( $attributes ) {
 		$link = esc_url( $link );
 		if ( $link ) {
-			$title = "<a href='{$link}'>{$title}</a>";
+			$processor = new WP_HTML_Tag_Processor( "<a href='{$link}'>{$title}</a>" );
+			$processor->next_tag( 'a' );
+
+			if ( $open_in_new_tab ) {
+				$processor->set_attribute( 'target', '_blank' );
+				$processor->set_attribute( 'rel', trim( $rel . ' noopener nofollow' ) );
+			} elseif ( '' !== $rel ) {
+				$processor->set_attribute( 'rel', $rel );
+			}
+
+			$title = "<div class='wp-block-rss__item-title'>" . $processor->get_updated_html() . "</div>";
```
@dhruvikpatel18, I'm sorry if my [previous comment](https://github.com/WordPress/gutenberg/pull/69641#pullrequestreview-2713151885) was unclear, but as I mentioned, we don't need to use HTML API here.

dhruvikpatel18 replied:
@Mamaduka Sorry for misunderstanding your previous feedback, Removed HTML API usage.

dhruvikpatel18 committed:
Removed HTML Tag Processor usage

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,15 +31,29 @@ function render_block_core_rss( $attributes ) {
+
+		if ( $open_in_new_tab ) {
+			$link_attributes .= ' target="_blank"';
+			$link_attributes .= ' rel="' . esc_attr( trim( $rel . ' noopener nofollow' ) ) . '"';
+		} elseif ( '' !== $rel ) {
+			$link_attributes .= ' rel="' . esc_attr( trim( $rel ) ) . '"';
```
I think we can drop `noopener`. See: https://github.com/WordPress/gutenberg/issues/26914#issuecomment-2612263053.

Additionally, `$rel` is already trimmed above, and another `trim` call seems unnecessary.

Mamaduka replied:
I also realized that this leaves `nofollow` hardcoded, and users won't be able to remove it, if the link opens in a new tab. Not 100% sure what the desired flow is here.

dhruvikpatel18 replied:
You're right about the hardcoded `nofollow`, actually it was there from previous implementation which i referred from social-link block.

Mamaduka commented:
@dhruvikpatel18, IIRC you can fix a failing JS unit test by running `npm run fixtures:regenerate`.

dhruvikpatel18 committed:
Allow users full control over the  attribute

dhruvikpatel18 committed:
Fixed JS Unit tests

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,15 +31,32 @@ function render_block_core_rss( $attributes ) {
 		$link = $item->get_link();
 		$link = esc_url( $link );
+
+		if ( $open_in_new_tab ) {
+			$link_attributes .= ' target="_blank"';
+
+			if ( '' !== $rel ) {
+				$link_attributes .= ' rel="' . esc_attr( $rel ) . '"';
+			}
+		} elseif ( '' !== $rel ) {
+			$link_attributes .= ' rel="' . esc_attr( $rel ) . '"';
+		}
```
Suggestion: The conditions here can be simplified.

```php
if ( $open_in_new_tab ) {
	$link_attributes .= ' target="_blank"';
}

if ( '' !== $rel ) {
	$link_attributes .= ' rel="' . esc_attr( $rel ) . '"';
}
```

dhruvikpatel18 committed:
Simplified the conditions for link attributes

Mamaduka commented on packages/block-library/src/rss/index.php:
```diff
@@ -31,15 +31,30 @@ function render_block_core_rss( $attributes ) {
+
+		if ( '' !== $rel ) {
+			$link_attributes .= ' rel="' . esc_attr( $rel ) . '"';
+		}
```
One last item. I think we can move this outside of the loop. There's no need to recreate `$link_attributes` on each iteration. It doesn't require any values from `$item`.

dhruvikpatel18 replied:
Yes, there is no need to  recreate `$link_attributes`, I've updated it.

Mamaduka reviewed (approved):
Thank you, @dhruvikpatel18!

The implementation looks good. There's one last nitpick, but generally, I think this is good to merge.

dhruvikpatel18 committed:
moved link atributes conditions outside the loop
